react-common/components/controls/TeachingBubble.tsx

```diff
@@ -0,0 +1,291 @@
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import { Button } from "./Button";
+import { ContainerProps, classList } from "../util";
+import { FocusTrap } from "./FocusTrap";
+import { useEffect } from "react";
+
+export enum Location {
+    Above,
+    Below,
+    Left,
+    Right,
+    Center
+}
+
+export interface TargetContent {
+    title: string;
+    description: string;
+    targetQuery: string;
+    location: Location;
+}
+
+export interface TeachingBubbleProps extends ContainerProps {
+    targetContent: TargetContent;
+    stepNumber: number;
+    totalSteps: number;
+    onClose: () => void;
+    parentElement?: Element;
+    activeTarget?: boolean; // if true, the target is clickable
+    onNext: () => void;
+    onBack: () => void;
+}
+
+export const TeachingBubble = (props: TeachingBubbleProps) => {
+    const {
+        id,
+        className,
+        ariaLabel,
+        ariaHidden,
+        ariaDescribedBy,
+        role,
+        targetContent,
+        onClose,
+        onNext,
+        onBack,
+        stepNumber,
+        totalSteps,
+        parentElement,
+        activeTarget
+    } = props;
+
+    const margin = 10;
+    const transparentBorder = `${margin}px solid transparent`;
+    const opaqueBorder = `${margin}px solid`;
+    let bubble: HTMLElement;
+    let bubbleArrow: HTMLElement;
+    let bubbleBounds: DOMRect;
+    let targetElement: HTMLElement;
+    let targetBounds: DOMRect;
+    let top: number;
+    let left: number;
+    let arrowTop: number;
+    let arrowLeft: number;
+
+    useEffect(() => {
+        positionBubbleAndCutout();
+        window.addEventListener("resize", positionBubbleAndCutout);
+        return () => {
+            window.removeEventListener("resize", positionBubbleAndCutout);
```

Comment:
I know the event listener is being added every time the stepNumber changes and I'm not sure whether this return will take care of cleaning all of them up.

I tried to add the event listener for resizing the window in a useEffect with an empty dependency array, but that resulted in the teaching bubbles and cutouts taking the shape of the first teaching bubble and cutout whenever the window was resized. It seems the event listener needs to be in this useEffect so that the position of the bubble and cutout are calculated with respect to the current step in the tour.

Please let me know if there might be an easier/better way to do this.

Reply:
Yes, that's the way to do this - useEffect runs whenever those values change (including trivially changing the first time it renders) & the intention is to have it rerun whenever things need to be recalculated (e.g. the dimensions of the region changing per step).

The return value here is used whenever it is 'cleaning up' before running the handler again or discarding the element, so yes, handlers should all end up cleaned up. If you wanna check this, you can see what event listeners are registered in the dev tools by clicking into the 'Event Listeners' tab and finding the event you're looking for:

![image](https://user-images.githubusercontent.com/5615930/225072079-5fb15e1e-e9b2-4ec6-ade1-f1df4eb54a1f.png)

which can be helpful if you're looking for memory / event listener leaks~